### DISCUSSION TITLE
Inter-Button communication API

### DISCUSSION THREAD

**Discussion by buffos** [elgatosf/elgato#149] [Ideas]
### Motivation

As is now, it is very hard to make **inter-button** communication. 
For example, I want to create an internal **OSC server** to listen to state changes on buttons and inform them (or **web server**) etc.

### How could it be implemented?

Following the current form of the API, we can add two functions. One to the StreamDeck and one to the Actions class to override:

`streamdeck.SendDataTo("keyString", dataObject)`  and  `action.Subscribe("keyString",obj, callbackfn)`

The form of Subscribe allows adding subscriptions to **many keys**.

### Usage

What this means is that actions will **subscribe** to these **key changes** and act if they are interested.
Keys can be URLs, OSC paths, names of applications, or anything else, and **inter-button communication** is very easy this way.

We could probably simulate that by setting values in global settings with a certain key and then deep linking to notify the action, which will read the global settings key and act accordingly, but this is too error-prone.

This will make it easier to make more general plugins.

For example, I have a service, and I publish new data every n seconds. I create my plugin, you set the interval rate, and the plugin publishes some keys. **Any plugin** can read and react to the values published.

**Comment by tiptronic** [comment]
hey @buffos 
likely I don't understand the request ( and honestly I don't know what OSC server is), but you can do all of this today using a bunch of ways, e.g:

- using a Proxy
- using Node's EventEmitter
- using a REST-API on  your own background-service
- using your own WebsocketServer (if you plan to talk (bi-directionally) to various plugins or even external apps).
- or even MQTT (or whatever) to talk to other machines, IOT etc..

Making this generally available on a plugin basis is likely a **call for trouble**, because then a plugin relies on (or needs) another plugin to add functionality

**Reply by buffos** [reply]
Is there any repo with examples of how to do that?
My only source of documentation is streamdeck's official documentation and the 3 examples on GitHub (especially the game example which was great). 

Can you please point to plugins using those methods (if it's possible)?

>Making this generally available on a plugin basis is likely a call for trouble, because then a plugin relies on (or needs) another plugin to add functionality

This is the downside, of course. The upside is deduplication. You do not have to implement the same functionality in many plugins. 
Besides, **we still rely** on other plugins to accomplish tasks (with **multi-actions**). The difference is that it is accomplished **by inclusion.**
What I was suggesting was **message-passing** inclusion.

multi-actions => cooperation by inclusion
message-passing => cooperation by messages

Thank you in advance for your time.

**Comment by tiptronic** [comment]
Hi @buffos 
there's a bunch of plugins out there which allow remote-connection, e.g.:

https://github.com/ybizeul/StreamDeckWS
e.g. in conjunction with NodeRed:
https://flows.nodered.org/node/node-red-contrib-streamdeck-ws

or with HomeAssistant:
https://github.com/cgiesche/streamdeck-homeassistant

or using MQTT:
https://marketplace.elgato.com/product/mqtt-99faa0b9-7d88-499a-ab4f-6d55f8c2e0a8

Just google for it and you will find many more...

If many plugins want to use the same functionality, then that's a good sign you need a background-service like our very own `Volume Controller plugin`:
https://marketplace.elgato.com/product/volume-controller-34d9aa59-a41a-4a4c-a853-202ca91409e1

https://www.youtube.com/watch?v=dDIl5CqJwRo

That way all plugins interested in communicaton with the service can install it and talk to it...

btw: Multi-Actions don't rely onto each other, but execute their actions one after another

**Reply by buffos** [reply]
Thank you. I will check them all out this weekend.

> background-service 

This is very close to what I have in mind.

I have just started coding for Stream Deck, and my main inspiration is the Elgato repository. It seems that the SDK has recently changed since most plugins are written in a slightly different way.

Anyway. My first attempt (in pre-alpha state) is https://github.com/buffos/streamdeck-json-group which lets you group and control buttons from a json file.

Thank you once again for your time.

**Reply by buffos** [reply]
One more question. 

All these examples use the old way of configuring the Stream Deck. Everything starts inside the `connectElgatoStreamDeckSocket(inPort, inPluginUUID, inRegisterEvent, inInfo)` function.

There we get access to the info about the websocket and create it to be used later.

If I go down that path, should I abandon the newer SDK approach? I tried adding that function with just getting access to the websocket, but everything stopped working.

**Reply by tiptronic** [reply]
@buffos No - of course not! I mentioned these example, because they reach out to external services.

The old and the new SDK have the same functionality, but the new SDK carries a lot of improvements (and also TypeScript typings, etc). 

I think you should first get aquainted with the (NEW) SDK itself. It has all the features the old SDK had - and more.

So once you understand the new SDK you will be able to 'translate' every old plugin to use the new SDK easily.

From a quick peek at your example plugin I als see you e.g. use `onPropertyInspectorDidDisappear` to act on settings change...
Why not `onDidReceiveSettings` where the plugin gets the settings as soon as the user changes it or when you call `setSettings`? 

I still don't know what all that OSC-things mean (or what OSC is), but to me it looks you're doing http-requests (?) through Powershell?
Isn't there an `official` API for OSC which you can use? In that cass you could use node's built-in mechanisms to simplify communication tremendously.

**Reply by buffos** [reply]
>Why not onDidReceiveSettings where the plugin

The reasoning behind that is, my guess is **onDidReceiveSettings** is called every time a parameter is changed. Since I do **file reads**, I wanted to optimize it and call it just once when we have no more properties to change. Closing the property inspector seemed a logical place to signal that no more property changes. (or I could add an apply button, but I thought it was unnecessary) 

>I still don't know what all that OSC-things mean

**OSC** is a "sound" protocol for communicating between devices, mainly over UDP. You can transmit number strings, etc., but it's fast and simple, since its binary, and it is used also in virtual production a lot to pass data quickly. Many applications support this kind of communication. much lighter than other protocols. Ideal for passing knob values, triggers, etc. 

What gRPC is to http requests is OSC to WebSockets (almost).

I tried using Node.js libraries for OSC to be able to do that, but for some reason they did not work, so I just wrote my own PowerShell function that does just that. Maybe it's because it's **mainly a UDP protocol**. and although I explicitly used that option in the node libraries, it did not work. Some OSC servers support OSC **over websockets**, but it's not universal.

>So once you understand the new SDK you will be able to 'translate' every old plugin to use the new SDK easily.

Definitely, but I could not find the equivalent of `connectElgatoStreamDeckSocket` to get the inPort. Perhaps I can find it in the `streamDeck.devices.onDeviceDidConnect` and the `DeviceDidConnectEvent` has the desired information. I will look at it.

**Reply by tiptronic** [reply]
> Definitely, but I could not find the equivalent of connectElgatoStreamDeckSocket to get the inPort. Perhaps I can find it in the streamDeck.devices.onDeviceDidConnect and the DeviceDidConnectEvent has the desired information. I will look at it.

You don't need that any more... it is handled by the SDK internally. Just call `streamdeck.connect()` (as you do in your plugin) and you're good to go... 

As far as `upd` is concerned, `node` has built-in UDP-support here:

`https://nodejs.org/api/dgram.html`

**Reply by tiptronic** [reply]
btw... a quick google-search unveiled this:

[https://www.npmjs.com/package/node-osc](https://www.npmjs.com/package/node-osc)

is that what you mean? (I have no OSC-server, so I can't test...)

**Reply by buffos** [reply]
But those plugins get access to the websocket to intercept messages. For example, the StreamDeckWS does the following inside the onmessage event listener.

```javascript
/*
		Every other message is simply forwarded to node-red, the condition is
		that the message must have a settings.remoteServer setting, which is the
		case for any key-related events.
		*/
		else if (jsonObj.hasOwnProperty("payload")) {
			if (jsonObj['payload'].hasOwnProperty("settings")) {
				key = jsonObj['payload']['settings']['remoteServer']
				if (isInMultiAction) {
					sendOneEvent(key,jsonObj)
				}
				else {
					if (connections.hasOwnProperty(key)) {
						c=connections[key].websocket
						if (c && c.readyState == 1) {
							console.log(jsonObj)
							c.send(JSON.stringify(jsonObj))
						}
					}
				}
			}
		}		
	};
```

How to do that in the new SDK. This is why I am asking about the websocket

**Reply by buffos** [reply]
> btw... a quick google-search unveiled this:
> 
> https://www.npmjs.com/package/node-osc
> 
> is that what you mean? (I have no OSC-server, so I can't test...)

Yes, I tested it, among others, and it did not work in any of the modes, and debugging that would take much more time than writing one myself.

**Reply by tiptronic** [reply]
@buffos 

> But those plugins get access to the websocket to intercept messages. For example, the StreamDeckWS does the following inside the onmessage event listener.

I don't know where you got the snippet, but it is certainly not sending back the event to Stream Deck using the plugin's websocket connection.
So you should first understand what the code does before hacking on it ;) 

If you don't know how to use websockets you should first read here:

[https://developer.mozilla.org/en-US/docs/Web/API/WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)

For me the code-snippet is out of context, but likely equivalent  to this:

```javascript
override onDidReceiveSettings(ev: DidReceiveSettingsEvent < ActionSettings >): Promise < void> | void {

	const settings = ev.payload.settings;  
	const isInMultiAction = ev.payload.isInMultiAction;
	// (or const { settings, isInMultiAction } = ev.payload;)
	const key = settings.remoteServer;
	if (isInMultiAction) {
		sendOneEvent(key, ev);
	} else {
		if (connections.hasOwnProperty(key)) {  // I don't know what connections is, but likely an array of some connected services/servers
			const c = connections[key].websocket;  // this has nothing to do with Stream Deck, but likely a websocket connection
			if (c && c.readyState === 1) {  // the websocket's state is open and ready
				console.log(ev);
				c.send(JSON.stringify(ev));  // send the payload to the websocket `c.send` is a method of the websocket object 
				// https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send
				
			}
		}
	}
};
```

...it would make no sense to send the ev back to Stream Deck at all...

**Reply by tiptronic** [reply]
If you want to just pass information from one action to another - you are aleady doing that properly in your sample:
`UpdateAllButtonDetails`

And if you want to communicate with other Makers, it's much more effective if you join us on discord:

https://discord.gg/StwbM2S9

**Reply by buffos** [reply]
@tiptronic.

>I don't know where you got the snippet, 

My question was not what the snippet does. I was hoping to capture custom "events" sent outside Stream Deck, inside the plugin. 
Probably this falls under the "WebSocket API", but there are no examples there.

Anyway. I probably wasted a lot of your time. I will try working around that with deep-links.

**Reply by tiptronic** [reply]
hey @buffos 
np - but using deep-links you usually talk from outside Stream Deck to Stream Deck.

If you want to talk (bi-directionally) to an outside service (or background-service or any kind of server, you usually use something like Websockets, Json-RPC, or whatever).

Here's a very quick overview of a websocket connection (this has **NOTHING** to do with Stream Deck and would also work in any browerer of plain Nodejs:


```javascript
const host = '127.0.0.1'; // address on your local machine
const port = 8987;  // some port on that address where the server listens on

const websocket = new WebSocket(`ws://${host}:${port}`); // now try to create a websocket connection

websocket.onopen = function() {
    // websocket opened
};

websocket.onclose = function(evt) {
    // websocket was closed
};

websocket.onerror = function(evt) {
    // error
};

websocket.onmessage = function(evt) {
   // message received
};

websocket.send('Hello, Server!');

```

In the `onmessage` event-handler you would get all messages from the server, which you then can dispatch to your plugin, action or do anything else inside your plugin.

That's it

**Reply by buffos** [reply]
@tiptronic 

Thank you once again.

>np - but using deep-links you usually talk from outside Stream Deck to Stream Deck.

yes. This is exactly what I want. outside Stream Deck initiated communication, not streamdeck initiated communication (which can be done in multiple ways, including ws)

I probably have some wrong assumptions. I thought that:

1. streamdeck internally has an open ws to a random port
2.  I could send data directly to that port from outside (but I do not know the port).
3. Receive the data as global settings and then do whatever I want.

So you propose instead to create internally a websocket to a predefined port and connect to the source(s) I want?

**Reply by tiptronic** [reply]
OK - that's easier to answer:

1 - NO
2 - NO
3 - NO

Last: if you don't want to open a security hole, the proper way would be to create an external service/background task and then connect from your plugin to it. (Much like requesting data using a `fetch` operation)
